### PR TITLE
Feature: event system and impression data events

### PR DESCRIPTION
**Pull request by daveleek**
# Description
Aims to add an event system to the SDK and a first implementation for impression data/events: https://docs.getunleash.io/advanced/impression-data#impression-event-data

Fixes # (issue)
#108 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?
- [x] ran all the existing unit tests
- [x] ran the cake script testing
- [x] added more unit tests and ran them
- [x] explicitly tested serialization/deserialization

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/Unleash/DefaultUnleash.cs [hunk 1]
```diff
@@ -235,6 +243,52 @@ private IEnumerable<Constraint> ResolveConstraints(ActivationStrategy activation
...
+                return;
+            }
+
+            try
+            {
+                var evtConfig = new EventCallbackConfig();
+                callback(evtConfig);
+                EventConfig = evtConfig;
+            }
+            catch (Exception ex)
```
Comment (sighphyre) [inline src/Unleash/DefaultUnleash.cs:260]: Is there something that can be raised here? This looks pretty harmless to me

File: src/Unleash/DefaultUnleash.cs [hunk 2]
```diff
@@ -235,6 +243,52 @@ private IEnumerable<Constraint> ResolveConstraints(ActivationStrategy activation
...
+                {
+                    Type = type,
+                    Context = context,
+                    EventId = Guid.NewGuid().ToString(),
+                    Enabled = enabled,
+                    FeatureName = name,
+                    Variant = variant
+                });
+            }
+            catch (Exception ex)
```
Comment (sighphyre) [inline src/Unleash/DefaultUnleash.cs:286]: Same as above, I can't see an obvious exception that could be caught here, is this just being defensive?

**Review by sighphyre — COMMENTED** [review]
This is pretty cool, just a small question about error handling but looks pretty darn good to me

File: src/Unleash/DefaultUnleash.cs (hunk 1 above)
Comment (daveleek) [reply src/Unleash/DefaultUnleash.cs:260]: This particular one is probably very harmless, it's the callback where you configure all the handlers for the various events. But anything started in here that fails on a file lock or whatever will crash the startup of Unleash

File: src/Unleash/DefaultUnleash.cs (hunk 2 above)
Comment (daveleek) [reply src/Unleash/DefaultUnleash.cs:286]: Here I'm trapping event emitting, if your logger fails for some reason for instance, that won't crash harder than an error in the log

**Comment by daveleek** [comment]
Thanks @sighphyre for review and questions. I've attempted a few more details in response. Perhaps this is a code area that could do with some comments 🤔 

I'm trapping exceptions here since these two callbacks are code coming from outside Unleash, but I'm open for anything :)

File: src/Unleash/DefaultUnleash.cs (hunk 1 above)
Comment (sighphyre) [reply src/Unleash/DefaultUnleash.cs:260]: I can stand by that!

File: src/Unleash/DefaultUnleash.cs (hunk 2 above)
Comment (sighphyre) [reply src/Unleash/DefaultUnleash.cs:286]: Sure, I can live with that, seems polite to log an error if something super unexpected happens here

**Review by sighphyre — APPROVED** [review]
Looks good to me

**Pull request merged**